BrainPortal/lib/boutiques_input_value_fixer.rb

```diff
@@ -0,0 +1,182 @@
+
+#
+# CBRAIN Project
+#
+# Copyright (C) 2008-2021
+# The Royal Institution for the Advancement of Learning
+# McGill University
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+# This module allows one to fix some of input parameters to specific values
+# The fixed input(s) would no longer be shown to the user in the form.
+#
+# In the descriptor, the spec would look like:
+#
+#    "custom": {
+#     "cbrain:integrator_modules": {
+#         "BoutiquesInputValueFixer": {
+#             "n_cpus": 1,
+#             "mem": "4G",
+#             "customquery": nil,
+#             "level": "group"
+#         }
+#     }
+# }
+# COULD BE TRICKY, PRESENTLY PARAMETER DEPENDENCY are not supported
+# disables-inputs, requires-inputs,
+# the parameters assigned null will be disabled without assignment
+# string 'null' is considered string for String params
+module BoutiquesInputValueFixer
+
+  # Note: to access the revision info of the module,
+  # you need to access the constant directly, the
+  # object method revision_info() won't work.
+  Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
+
+  # the hash of param values to be fixed or be ommited ()  #
+  def fixation
+    fixation = self.boutiques_descriptor.custom_module_info('BoutiquesInputValueFixer')
+    cb_error "module BoutiquesInputValueFixer requires a hash" unless fixation.is_a? Hash
+    fixation
+  end
+
+  # deletes fixed inputs listed in the custom 'integrator_modules'
+  # no input or values dependencies for fixed variables are fully supported,
+  # in the presence of dependencies involving fixed params, the module
+  # does its best to avoid deadlock and issues, but, probably,
+  # might fails in edge cases. It is recommended to supply fixation is "closed" in the sense that
+  # contains all the implied parameter fixes or deletions
+  def delete_fixed_inputs(descriptor)
+    # input parameters are marked by null values will be excluded from the command line
+    # the major use case are Flags, but also may be useful to address params with 'default' (
+    # or, for flags, null-like values)
+
+    descriptor_dup = descriptor.dup
+    skipped = fixation.keys.select do |i_id|
```

Comment:
What is `skipped` exactly? I see a big select loop but there is no explanation as to what this contains at the end, exactly. It seems to be a subset of the IDs of the fixed inputs? But which subset, and what is it for?

Reply:
maybe `to_be_deleted` ? `to_abstract_away` ?